### PR TITLE
[SPARK-17034][SQL] adds expression UnresolvedOrdinal to represent the ordinals in GROUP BY or ORDER BY

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

This PR adds expression `UnresolvedOrdinal` to represent the ordinal in GROUP BY or ORDER BY, and fixes the rules when resolving ordinals.

Ordinals in GROUP BY or ORDER BY like `1` in `order by 1` or `group by 1` should be considered as unresolved before analysis. But in current code, it uses `Literal` expression to store the ordinal. This is inappropriate as `Literal` itself is a resolved expression, it gives the user a wrong message that the ordinals has already been resolved. 
### Before this change

Ordinal is stored as `Literal` expression

```
scala> sc.setLogLevel("TRACE")
scala> sql("select a from t group by 1 order by 1")
...
'Sort [1 ASC], true  
 +- 'Aggregate [1], ['a]
     +- 'UnresolvedRelation `t
```

For query:

```
scala> Seq(1).toDF("a").createOrReplaceTempView("t")
scala> sql("select count(a), a from t group by 2 having a > 0").show
```

During analysis, the intermediate plan before applying rule `ResolveAggregateFunctions` is:

```
'Filter ('a > 0)
   +- Aggregate [2], [count(1) AS count(1)#83L, a#81]
        +- LocalRelation [value#7 AS a#9]
```

Before this PR, rule `ResolveAggregateFunctions` believes all expressions of `Aggregate` have already been resolved, and tries to resolve the expressions in `Filter` directly. But this is wrong, as ordinal `2` in Aggregate is not really resolved!
### After this change

Ordinals are stored as `UnresolvedOrdinal`.

```
scala> sc.setLogLevel("TRACE")
scala> sql("select a from t group by 1 order by 1")
...
'Sort [unresolvedordinal(1) ASC], true
 +- 'Aggregate [unresolvedordinal(1)], ['a]
      +- 'UnresolvedRelation `t`
```
## How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by clockfly** [comment]
Ping @cloud-fan

**Comment by rxin** [comment]
Is there a way to fix this without introducing the new unresolved ordinal classes? If there isn't, can we combine the two into a single UnresolvedOrdinal?

**Comment by clockfly** [comment]
@rxin 
I have to introdce a new unresolved expression as "group by ordinal" and "order by ordinal" can be turned off by a config like "conf.groupByOrdinal" or "conf.orderByOrdinal".

Yes, `UnresolvedOrdinal` is a good idea.

**Comment by cloud-fan** [comment]
Can we just add a check in `ResolveAggregateFunctions`? If `groupByOrdinal` is enabled and aggregate expressions contains integral literals, we skip it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -2223,3 +2223,29 @@ object TimeWindowing extends Rule[LogicalPlan] {
       }
   }
 }
+
+/**
+ * Replaces ordinal in 'order by' or 'group by' with unresolved UnresolvedOrdinal expression.
+ */
+class UnresolvedOrdinalSubstitution(conf: CatalystConf) extends Rule[LogicalPlan] {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: if we end up doing it this way, move this to its own file, and create an invididual test suite. 

analyzer file is getting too large.

File: sql/core/src/test/resources/sql-tests/results/group-by-ordinal.sql.out [hunk 2]
```diff
@@ -95,7 +95,7 @@ select a, b from data group by -1
 struct<>
 -- !query 8 output
 org.apache.spark.sql.AnalysisException
-GROUP BY position -1 is not in select list (valid range is [1, 2]); line 1 pos 31
+GROUP BY position -1 is not in select list (valid range is [1, 2]); line 1 pos 22
```
Comment (rxin) [inline sql/core/src/test/resources/sql-tests/results/group-by-ordinal.sql.out]: why does the position change?

Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/group-by-ordinal.sql.out]: the pos is wrong now

**Comment by clockfly** [comment]
@cloud-fan 
ResolveAggregateFunctions is only one reference of `Aggregate`, there can be other references? If not now, will we add some rule to reference Aggregate in future? The case here is that `Aggregate` itself is unresolved when it contains group by `ordinal`. The current fix is fail-proof for all potential usages.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Ok.

File: sql/core/src/test/resources/sql-tests/results/group-by-ordinal.sql.out (hunk 2 above)
Comment (clockfly) [inline sql/core/src/test/resources/sql-tests/results/group-by-ordinal.sql.out]: Sorry, I will fix this. I didn't understand the meaning of pos before.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -2223,3 +2223,29 @@ object TimeWindowing extends Rule[LogicalPlan] {
...
+  private def isIntegerLiteral(sorter: Expression) = IntegerIndex.unapply(sorter).nonEmpty
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case s @ Sort(orders, _, _) if conf.orderByOrdinal &&
+      orders.exists(o => isIntegerLiteral(o.child)) =>
+      val newOrders = orders.map {
+        case order @ SortOrder(IntegerIndex(index), _) =>
+          order.copy(child = UnresolvedOrdinal(index))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: we need a way to move the line position information.

**Comment by cloud-fan** [comment]
The bug described in JIRA is already fixed by https://github.com/apache/spark/pull/14595 by accident...

shall we open a new JIRA as this PR is actually a refactor/cleanup?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -32,7 +32,7 @@ import org.apache.spark.sql.catalyst.planning.IntegerIndex
 import org.apache.spark.sql.catalyst.plans._
 import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, _}
 import org.apache.spark.sql.catalyst.rules._
-import org.apache.spark.sql.catalyst.trees.TreeNodeRef
+import org.apache.spark.sql.catalyst.trees.{CurrentOrigin, Origin, TreeNodeRef}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: no need to import here?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 5]
```diff
@@ -377,4 +377,43 @@ class AnalysisSuite extends AnalysisTest {
...
+    val b = testRelation2.output(1)
+    val conf = new SimpleCatalystConf(caseSensitiveAnalysis = true)
+
+    // Expression OrderByOrdinal is unresolved.
+    assert(!UnresolvedOrdinal(0).resolved)
+
+    // Tests order by ordinal, apply single rule.
+    val plan = testRelation2.orderBy(Literal(1).asc, Literal(2).asc)
+    comparePlans(
+      new UnresolvedOrdinalSubstitution(conf).apply(plan),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: actually we don't have unit test for single analyzer rule yet, this suite is for the whole analyzer. We should write test like
```
val query = testRelation2.orderBy(Literal(1).asc, Literal(2).asc)
val expected = testRelation2.orderBy('a.asc, 'b.asc)
checkAnalysis(query, expected)
```

Or we can create a new suite and begin to test single analyzer rule

Comment (clockfly) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: I will creates a new test suite.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**